h2o-core/src/main/java/water/rapids/BinaryMerge.java

```diff
@@ -357,145 +351,147 @@ private long bsearchLeft(long x, boolean returnLow, long upp)  {
     return returnLow ? low : upp;
   }
 
-  // Must be passed two leftKeys only.
-  // Optimized special case for the two calling points; see usages in bmerge_r below.
-  private boolean leftKeyEqual(byte x[][], long xi, long yi) {
-    byte xbatch[] = x[(int)(xi / _leftKO._batchSize)];
-    byte ybatch[] = x[(int)(yi / _leftKO._batchSize)];
-    int xoff = (int)(xi % _leftKO._batchSize) * _leftSB._keySize;
-    int yoff = (int)(yi % _leftKO._batchSize) * _leftSB._keySize;
-    int i=0;
-    while (i<_leftSB._keySize && xbatch[xoff++] == ybatch[yoff++]) i++;
-    return(i==_leftSB._keySize);
-  }
+  /***
+   * For a specific MSB, if we can find MSBs in the left frame and rite frame that contains values in the MSB range,
+   * we will try to find match between the left frame and rite frame and include those rows in the final merged frame.
+   * If no match is found between the two frames but allLeft = true, we will go ahead and just include all the rows
+   * having the same MSB in the left frame into the final merged frame.
+   * 
+   * If allLeft is true, we will always iterate over the leftFrame keys and use binary search over the rite frame keys.
+   * We will count all duplicate keys in the left and rite frames as well.  During the search process, we should be 
+   * able to shrink the range of rite frame keys to search over as we gain more knowledge over the relative key 
+   * sizes between the two frames and all keys are sorted already.
+   * 
+   * However, when allLeft is false, we will iterate over the frames that has the smallest number of keys instead. 
+   * Hence, in this case, we can search over either the left or rite frame depending on the number of rows to be 
+   * searched through.  
+   * 
+   * @param leftLowIn: left frame lowest row number minus 1
+   * @param leftUppIn: number of rows in left frame with specific MSB
+   * @param riteLowIn: rite frame lowest row number minus 1 with specific MSB
+   * @param riteUppIn: number of rows in rite frame with specific MSB
+   *                  
+   */
+  private void binaryMergeOneMSB(long leftLowIn, long leftUppIn, long riteLowIn, long riteUppIn) {
+    if (!_allLeft && riteUppIn == 0) return;  // no merging possible with empty rite frame here
+    boolean leftFrameIterate = _allLeft ? true : ((leftUppIn - leftLowIn) > (riteUppIn - riteLowIn) ? false : true);
```

Comment:
leftFrameIterate can definitely be final, so can iterUppIn, maybe other as well?

Reply:
yes, I added them.